File: 2022/hs/day-21.hs
import Data.Map (Map)
import qualified Data.Map as Map
import Text.Read (readMaybe)
main = do
    input <- readFile "inputs/21.txt"
    let yells = asYells input
    print $ getYell "root" yells
    print $ (solve . uncurry intoEquation . initPt2) yells

-- part A
type PendingCalc = (String, Operation, String)
data Operation = Plus | Minus | Mul | Div deriving Eq

data Yells = Yells {
    resolved :: Map String Int,
    awaiting :: Map String PendingCalc
} deriving Eq

asYells s = foldr applyLine (Yells Map.empty Map.empty) (lines s)
    where
        parseCalc s = let (l, ' ':op:' ':r) = break (== ' ') s in
            case op of
                '+' -> (l, Plus, r)
                '-' -> (l, Minus, r)
                '*' -> (l, Mul, r)
                '/' -> (l, Div, r)
                _ -> error "unimplemented"
        applyLine line (Yells resolved awaiting) = let (name, ':':' ':calc) = splitAt 4 line in
            case readMaybe calc of
                Just i -> Yells (Map.insert name i resolved) awaiting
                Nothing -> Yells resolved (Map.insert name (parseCalc calc) awaiting)

apply Plus = (+)
apply Minus = (-)
apply Mul = (*)
apply Div = div

resolveOnce (Yells resolved awaiting) = Yells (Map.union resolved resolvables) (Map.difference awaiting resolvables)
    where 
        resolve (left, op, right) = do
            lval <- Map.lookup left resolved
            rval <- Map.lookup right resolved
            return $ apply op lval rval
        resolvables = Map.mapMaybe resolve awaiting
getYell s y @ (Yells resolved _) = case Map.lookup s resolved of
    Just i -> i
    Nothing -> getYell s (resolveOnce y)

-- part B
initPt2 (Yells resolved awaiting) = (
        fixed resolveOnce (Yells (Map.delete "humn" resolved) (Map.delete "root" awaiting)), 
        let (l, _, r) = awaiting Map.! "root" in (l, r)
    )
    where
        fixed f a
            | a == f a = a
            | otherwise = fixed f (f a)

intoEquation (Yells resolved awaiting) (r0, r1) = Equation (parseLHS lhs) rhs
    where
        value t = Map.lookup t resolved
        (lhs, rhs) = case (value r0, value r1) of
            (Just lval, _) -> (r1, lval)
            (_, Just rval) -> (r0, rval)
            _ -> error "expected resolution"
        
        parseLHS "humn" = Var
        parseLHS s = let (l, op, r) = awaiting Map.! s in
            case (value l, value r) of
                (Just lval, _) -> KLeft lval op (parseLHS r)
                (_, Just rval) -> KRight (parseLHS l) op rval
                _ -> error "expected resolution"

data LHS = KLeft Int Operation LHS | KRight LHS Operation Int | Var
data Equation = Equation LHS Int

invert (KLeft a Plus _) = subtract a
invert (KLeft a Minus _) = (a - )
invert (KLeft a Mul _) = (`div` a)
invert (KLeft a Div _) = (a `div`)

invert (KRight _ Plus b) = subtract b
invert (KRight _ Minus b) = (+ b)
invert (KRight _ Mul b) = (`div` b)
invert (KRight _ Div b) = (* b)

invert Var = id

solve (Equation lhs @ (KLeft _ _ lhs') rhs) = solve (Equation lhs' (invert lhs rhs))
solve (Equation lhs @ (KRight lhs' _ _) rhs) = solve (Equation lhs' (invert lhs rhs))
solve (Equation Var rhs) = rhs

File: 2022/src/bin/day-21.rs
use std::collections::HashMap;
use std::fs;
use std::str::FromStr;

fn main() {
    let input = fs::read_to_string("inputs/21.txt").unwrap();
    
    let mut yells: Yells = input.parse().unwrap();
    println!("{:?}", yells.get("root"));

    let humn = input.parse::<Yells>()
        .unwrap()
        .pt2();
    println!("{:?}", humn);
}

#[derive(Debug)]
enum Operation { Plus, Minus, Mul, Div }

#[derive(Debug)]
struct PendingCalc {
    left: String,
    right: String,
    op: Operation
}

impl PendingCalc {
    fn try_apply(&self, resolved: &HashMap<String, isize>) -> Option<isize> {
        let &left = resolved.get(&self.left)?;
        let &right = resolved.get(&self.right)?;

        match self.op {
            Operation::Plus  => Some(left + right),
            Operation::Minus => Some(left - right),
            Operation::Mul   => Some(left * right),
            Operation::Div   => Some(left / right),
        }
    }
}

#[derive(Debug)]
struct Yells {
    resolved: HashMap<String, isize>,
    awaiting: HashMap<String, PendingCalc>
}

impl Yells {
    fn new() -> Self {
        Self {
            resolved: HashMap::new(),
            awaiting: HashMap::new()
        }
    }

    fn resolve_awaits(&mut self) {
        self.awaiting.retain(|k, v| {
            v.try_apply(&self.resolved).map(|i| {
                self.resolved.insert(k.clone(), i);
            }).is_none()
        });
    }

    fn get(&mut self, k: &str) -> Option<isize> {
        while !self.resolved.contains_key(k) && !self.awaiting.is_empty() {
            self.resolve_awaits();
        }
        self.resolved.get(k).copied()
    }

    fn pt2(mut self) -> isize {
        self.resolved.remove("humn");
        let root = self.awaiting.remove("root").unwrap();

        let mut awaiting = self.awaiting.len();
        loop {
            self.resolve_awaits();
            if awaiting != self.awaiting.len() {
                awaiting = self.awaiting.len();
            } else {
                break;
            }
        }
        
        let resolved = self.resolved;
        let mut awaiting = self.awaiting;
        let (r0, r1) = (root.left, root.right);

        // create equation:
        // pcs are the operations applied to get from humn -> root
        // these are right to left, so imagine folding from the right to get from humn to root
        // it's awful i know but simplifies some code
        
        // rhs is the value needed for root
        let mut pcs = vec![];

        let (mut lhs, rhs) = if let Some(&lval) = resolved.get(&r0) {
            (r1, lval)
        } else if let Some(&rval) = resolved.get(&r1) {
            (r0, rval)
        } else {
            panic!("Both sides of root are unresolved {:?} :(", (r0, r1))
        };

        let mut pc;
        while &lhs != "humn" {
            let calc = awaiting.remove(&lhs).unwrap();
            (pc, lhs) = PartialCalc::resolve(calc, &resolved);
            pcs.push(pc);
        }

        pcs.into_iter()
            .fold(rhs, |rhs, op| op.inverse(rhs))
    }
}

impl FromStr for Yells {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut y = Yells::new();
        for line in s.lines() {
            let (name, data) = line.split_once(": ").ok_or(())?;
            if let Ok(calc) = data.parse() {
                y.awaiting.insert(String::from(name), calc);
            } else if let Ok(num) = data.parse() {
                y.resolved.insert(String::from(name), num);
            } else {
                Err(())?
            }
        }

        Ok(y)
    }
}
impl FromStr for PendingCalc {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut it = s.split(' ');
        let sleft = it.next().ok_or(())?;
        let sop = it.next().ok_or(())?;
        let sright = it.next().ok_or(())?;

        let op = match sop {
            "+" => Operation::Plus,
            "-" => Operation::Minus,
            "*" => Operation::Mul,
            "/" => Operation::Div,
            _ => unimplemented!()
        };

        Ok(Self {
            left: String::from(sleft),
            right: String::from(sright),
            op,
        })
    }
}

#[derive(Debug)]
enum PartialCalc {
    Left(isize, Operation),
    Right(Operation, isize)
}
impl PartialCalc {
    fn resolve(calc: PendingCalc, resolved: &HashMap<String, isize>) -> (Self, String) {
        let PendingCalc { left, right, op } = calc;

        if let Some(&lval) = resolved.get(&left) {
            (PartialCalc::Left(lval, op), right)
        } else if let Some(&rval) = resolved.get(&right) {
            (PartialCalc::Right(op, rval), left)
        } else {
            panic!("Both sides are unresolved ({left:?} {op:?} {right:?}) :(")
        }
    }

    fn inverse(&self, rhs: isize) -> isize {
        match self {
            PartialCalc::Left(a, op) => match op {
                Operation::Plus => rhs - a,
                Operation::Minus => a - rhs,
                Operation::Mul => rhs / a,
                Operation::Div => a / rhs,
            },
            PartialCalc::Right(op, b) => match op {
                Operation::Plus => rhs - b,
                Operation::Minus => rhs + b,
                Operation::Mul => rhs / b,
                Operation::Div => rhs * b,
            },
        }
    }
}